File: Problems/19. Remove Nth Node From End of List/solution.cpp
#include <iostream>

using namespace std;

 struct ListNode {
     int val;
     ListNode *next;
     ListNode() : val(0), next(nullptr) {}
     ListNode(int x) : val(x), next(nullptr) {}
     ListNode(int x, ListNode *next) : val(x), next(next) {}
 };

class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        if (head->next == nullptr) return nullptr;

        ListNode* curr = head;
        int no = 0;
        while (curr != nullptr) {
            no++;
            curr = curr->next;
        }

        n = no-n;
        if (n == 0) head = head->next;
        else {
            curr = head;
            while (n > 1) {
                n--;
                curr = curr->next;
            }
            curr->next = curr->next->next;
        }

        return head;
    }
};

int main () {

    return 0;
}

File: Problems/19. Remove Nth Node From End of List/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        if not head:
            return head
        
        pos = {}

        cur = head
        nr = 0
        while cur:
            pos[nr] = cur
            nr+=1
            cur = cur.next

        del_pos = nr - n - 1

        if del_pos == -1:
            head = head.next
            return head

        node = pos[del_pos]
        node.next = node.next.next if node.next else None

        return head
        
        